controllers/services_controller.go

```diff
@@ -0,0 +1,184 @@
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	MetricsServiceName = "ssp-operator-metrics"
+	OperatorName       = "ssp-operator"
+)
+
+func ServiceObject(namespace string) *v1.Service {
+	policyCluster := v1.ServiceInternalTrafficPolicyCluster
+	familyPolicy := v1.IPFamilyPolicySingleStack
+	return &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      MetricsServiceName,
+			Namespace: namespace,
+			Labels: map[string]string{
+				metrics.PrometheusLabelKey: metrics.PrometheusLabelValue,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			InternalTrafficPolicy: &policyCluster,
+			IPFamilies:            []v1.IPFamily{v1.IPv4Protocol},
+			IPFamilyPolicy:        &familyPolicy,
+			Ports: []v1.ServicePort{
+				{
+					Name:       metrics.MetricsPortName,
+					Port:       443,
+					Protocol:   v1.ProtocolTCP,
+					TargetPort: intstr.FromString(metrics.MetricsPortName),
+				},
+			},
+			Selector: map[string]string{
+				metrics.PrometheusLabelKey: metrics.PrometheusLabelValue,
+				"name":                     OperatorName,
+			},
+			SessionAffinity: v1.ServiceAffinityNone,
+			Type:            v1.ServiceTypeClusterIP,
+		},
+	}
+}
+
+// Annotation to generate RBAC roles to read and modify services
+// +kubebuilder:rbac:groups="",resources=services,verbs=get;watch;list;create;update;delete
+
+func CreateServiceController(mgr ctrl.Manager) (*serviceReconciler, error) {
+	return newServiceReconciler(mgr)
+}
+
+func (r *serviceReconciler) Start(ctx context.Context, mgr ctrl.Manager) error {
+	err := r.createMetricsService(ctx)
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return fmt.Errorf("error start serviceReconciler: %w", err)
+	}
+
+	return r.setupController(mgr)
+}
+
+func (r *serviceReconciler) setServiceOwnerReference(service *v1.Service) error {
+	return controllerutil.SetOwnerReference(r.deployment, service, r.client.Scheme())
+}
+
+func (r *serviceReconciler) createMetricsService(ctx context.Context) error {
+	service := ServiceObject(r.serviceNamespace)
+	err := r.setServiceOwnerReference(service)
+	if err != nil {
+		return fmt.Errorf("error setting owner reference: %w", err)
+	}
+	return r.client.Create(ctx, service)
+}
+
+func (r *serviceReconciler) setupController(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("service-controller").
+		For(&v1.Service{}, builder.WithPredicates(predicate.NewPredicateFuncs(
+			func(object client.Object) bool {
+				return object.GetName() == MetricsServiceName && object.GetNamespace() == r.serviceNamespace
+			}))).
+		Complete(r)
+}
+
+// serviceReconciler reconciles the required services in the operator's namespace
+type serviceReconciler struct {
+	client           client.Client
+	log              logr.Logger
+	serviceNamespace string
+	deployment       *apps.Deployment
+}
+
+func setDeploymentField(r *serviceReconciler, apiReader client.Reader) error {
```

Comment:
I meant the function could look like:
```go
func getOperatorDeployment(namespace string, apiReader client.Reader) (*apps.Deployment, error) {
	objKey := client.ObjectKey{Namespace: namespace, Name: OperatorName}
	var deployment apps.Deployment
	err := apiReader.Get(context.TODO(), objKey, &deployment)
	if err != nil {
		return nil, fmt.Errorf("setDeploymentField, get deployment: %w", err)
	}
	return &deployment, nil
}
```

And then it would be called like this:
```go
func newServiceReconciler(mgr ctrl.Manager) (*serviceReconciler, error) {
	logger := ctrl.Log.WithName("controllers").WithName("Resources")
	namespace, err := getOperatorNamespace(logger)
	if err != nil {
		return nil, err
	}

	deployment, err := getOperatorDeployment(namespace, mgr.GetAPIReader())
	if err != nil {
		return nil, err
	}

	reconciler := &serviceReconciler{
		client:           mgr.GetClient(),
		log:              logger,
		serviceNamespace: namespace,
		deployment:       deployment,
	}

	return reconciler, nil
}
```

Reply:
ok